File: leetcode/medium/max_ramp.py



def maxRamp(arr):
    width = 0
    minIndex = len(arr) * 2;
    for i in sorted(range(len(arr)), key = arr.__getitem__):
        width = max(width, i - minIndex)
        minIndex = min(i, minIndex)  
    return width

print(maxRamp([9,8,1,0,1,9,4,0,4,1]))

File: leetcode/medium/max_ramp.cc
#include <algorithm>
#include <vector>

#define min(a,b) ( a < b ? a : b)
#define max(a,b) ( a > b ? a : b)


using namespace std; 

// leetcode: 

int maxRamp(vector<int>& arr) {
    int maxWidth = 0;
    int minIndex = arr.size() + 1;
    vector<pair<int,int>> indexes;
    for(int i = 0; i < arr.size(); ++i) {
       indexes.push_back(pair<int,int>(arr[i],i)); 
    }
    sort(indexes.begin(), indexes.end(), [](pair<int,int>& a, pair<int,int>& b) {
        return a.first < b.first;
    });

    for(int i = 0; i < indexes.size(); ++i) {
        maxWidth = max(maxWidth, indexes[i].second - minIndex);
        minIndex = min(minIndex, indexes[i].second); 
    }
    return maxWidth;
}
